src/crimson/os/seastore/journal.cc

```diff
@@ -0,0 +1,365 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include <boost/iterator/counting_iterator.hpp>
+
+#include "crimson/os/seastore/journal.h"
+
+#include "include/intarith.h"
+#include "crimson/os/seastore/segment_manager.h"
+
+namespace {
+  seastar::logger& logger() {
+    return crimson::get_logger(ceph_subsys_filestore);
+  }
+}
+
+namespace crimson::os::seastore {
+
+Journal::Journal(SegmentManager &segment_manager)
+  : block_size(segment_manager.get_block_size()),
+    max_record_length(
+      segment_manager.get_segment_size() -
+      p2align(ceph::encoded_sizeof_bounded<segment_header_t>(),
+	      size_t(block_size))),
+    segment_manager(segment_manager) {}
+
+
+Journal::initialize_segment_ertr::future<> Journal::initialize_segment(
+  Segment &segment)
+{
+  logger().debug("initialize_segment {}", segment.get_segment_id());
+  // write out header
+  ceph_assert(segment.get_write_ptr() == 0);
+  bufferlist bl;
+  auto header = segment_header_t{
+    current_journal_segment_seq++,
+    segment.get_segment_id(),
+    current_replay_point};
+  ::encode(header, bl);
+
+  written_to = segment_manager.get_block_size();
+  return segment.write(0, bl).handle_error(
+    init_ertr::pass_further{},
```

Comment:
current_journal_segment_seq--; ?

Reply:
Unlikely, we probably don't want to retry the same segment if we failed to write.  Mostly, however, the machinery for reusing segments basically doesn't exist yet, so this will all get a rewrite pass in the future.  As such, I don't want to spend a lot of effort on it yet.